src/main/java/de/tum/in/www1/artemis/service/LearningGoalProgressService.java

```
@@ -0,0 +1,189 @@
+        this.exerciseRepository = exerciseRepository;
+        this.lectureUnitRepository = lectureUnitRepository;
+    }
+
+    /**
+     * Asynchronously update the progress for all learning goals linked to the given learning object
+     * @param learningObject The learning object for which to fetch the learning goals
+     * @param participant The participant (user or team) for which to update the progress
+     */
+    @Async
+    public void updateProgressByLearningObjectAsync(ILearningObject learningObject, @NotNull Participant participant) {
+        updateProgressByLearningObject(learningObject, participant);
+    }
+
+    /**
+     * Update the progress for all learning goals linked to the given learning object
+     * @param learningObject The learning object for which to fetch the learning goals
+     * @param participant The participant (user or team) for which to update the progress
+     */
+    public void updateProgressByLearningObject(ILearningObject learningObject, @NotNull Participant participant) {
+        logger.debug("Updating learning goal progress for participant {}.", participant.getName());
+        try {
+            SecurityUtils.setAuthorizationObject();
+
+            Set<LearningGoal> learningGoals;
+            if (learningObject instanceof Exercise exercise) {
+                learningGoals = exerciseRepository.findByIdWithLearningGoals(exercise.getId()).map(Exercise::getLearningGoals).orElse(null);
+            }
+            else if (learningObject instanceof LectureUnit lectureUnit) {
+                learningGoals = lectureUnitRepository.findByIdWithLearningGoals(lectureUnit.getId()).map(LectureUnit::getLearningGoals).orElse(null);
+            }
+            else {
+                return;
+            }
+
+            if (learningGoals == null) {
+                // Learning goals couldn't be loaded, the exercise/lecture unit might have already been deleted
+                return;
```

Would be interesting for debugging purposes to log a message here instead of completely silently returning?